---
title: "Feasibility report"
subtitle:  "based on email from Joaquin Mateo"
author: "Alex Paynter"
date: "`r Sys.Date()`"
output: 
  html_document:
    toc: true
    toc_float: true
    theme: paper
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = F, include = F,
                      warning = F, message = F,
                      fig.width = 7, fig.height = 5)
k_dpi <- 150
pal_pt_hc <- c("#004488","#ddaa33", "#bb5566")

# Todo before sending:
# 1. check with Jen
# 2. check logic in drug overlap function with the work you did in the off label paper.
```

<style type="text/css">
.main-container {
  max-width: 1000px !important;
  margin: auto;
}
</style>

```{r}
library(purrr)
library(here)
library(fs)

purrr::walk(.x = fs::dir_ls(here('R')), .f = source) # also loads lots of packages.


```



```{r, set_gtsummary_theme}
theme_gtsummary_compact()
theme_gtsummary_language("en", big.mark = "") # for now.


```


```{r, warning = T}
read_wrap_clin <- function(p) {
  read_rds(file = here("data", 'clin', p))
}

dft_pt <- read_wrap_clin("dft_pt.rds")
dft_ca_ind <- read_wrap_clin("dft_ca_ind.rds")
# we're taking the augmented version, which has TMB columns added.  The existing info is the same.
dft_cpt <- read_wrap_clin("dft_cpt_aug.rds")
dft_reg <- read_wrap_clin("dft_reg.rds")
dft_med_onc <- read_wrap_clin('dft_med_onc.rds')
dft_path <- read_wrap_clin('dft_path.rds')
dft_drug <- read_csv(
  file = here('data', 'drug.csv')
)
```


```{r, warning = F}
dft_drug %<>% 
  mutate(
    agent = str_replace_all(drug, "\\(.*", "")
  ) %>%
  relocate(
    agent,
    .before = drug
  )

vec_agent_nha <- c(
  'Abiraterone Acetate',
  'Enzalutamide',
  'Darolutamide',
  'Apalutamide'
)

dft_drug_nha <- dft_drug %>%
  filter(agent %in% vec_agent_nha)

dft_drug_nha_first_last <- dft_drug_nha %>% 
  group_by(record_id, ca_seq) %>%
  summarize(
    nha_used = paste(sort(unique(agent)), collapse = ", "),
    first_nha_yrs = min(dx_drug_start_int, na.rm = T) / 365.25,
    last_nha_yrs = max(dx_drug_end_or_lastadm_int, na.rm = T) / 365.25,
    .groups = 'drop'
  )
```


```{r}
dft_dmet_time <- make_dmet_status_block(ca_ind_dat = dft_ca_ind) %>%
  filter(dmet_status %in% 'Distant Metastasis') %>%
  rename(dx_dmet_yrs = dx_block_start) %>%
  select(record_id, ca_seq, dx_dmet_yrs)

dft_crpc_time <- make_cast_status_block(
  dft_med_onc,
  dft_ca_ind
) %>%
  filter(md_cast_status_f %in% 'Castrate-Resistant') %>%
  rename(dx_crpc_yrs = dx_block_start) %>%
  select(record_id, dx_crpc_yrs)

dft_first_cpt <- full_join(
  get_first_cpt(dft_ca_ind, dft_cpt, type = 'order'),
  get_first_cpt(dft_ca_ind, dft_cpt, type = 'report'),
  by = c('record_id', 'ca_seq')
) %>%
  mutate(
    dx_cpt_best_yrs = if_else(
      is.na(dx_cpt_order_yrs),
      dx_cpt_rep_yrs,
      dx_cpt_order_yrs
    )
  )
  
  

dft_joined_times <- full_join(
  dft_drug_nha_first_last,
  dft_dmet_time,
  by = c("record_id", "ca_seq"),
  relationship = 'one-to-one'
) %>%
  full_join(
    dft_first_cpt,
    by = c("record_id", 'ca_seq'),
    relationship = 'one-to-one'
  ) %>%
  full_join(
    dft_crpc_time,
    by = 'record_id',
    relationship = 'one-to-one'
  )
```


```{r}
n_cohort <- nrow(dft_ca_ind)

n_nha <- dft_joined_times %>%
  filter(!is.na(first_nha_yrs)) %>%
  nrow(.)

n_mcrpc <- dft_joined_times %>%
  filter(!is.na(dx_dmet_yrs) & !is.na(dx_crpc_yrs)) %>%
  nrow(.)

n_nha_anytime_mcrpc <- dft_joined_times %>%
  filter(!is.na(dx_dmet_yrs) & 
           !is.na(dx_crpc_yrs) & 
           !is.na(first_nha_yrs)) %>%
  nrow(.)

n_nha_post_mcrpc <- dft_joined_times %>%
  filter(
    !is.na(dx_dmet_yrs) & 
      !is.na(dx_crpc_yrs) & 
      !is.na(first_nha_yrs)
  ) %>%
  filter(last_nha_yrs > dx_dmet_yrs) %>%
  filter(last_nha_yrs > dx_crpc_yrs) %>%
  nrow(.)

n_nha_post_mcrpc_seq <- dft_joined_times %>%
  filter(!is.na(dx_dmet_yrs) & !is.na(dx_crpc_yrs) & !is.na(first_nha_yrs)) %>%
  filter(last_nha_yrs > dx_dmet_yrs) %>%
  filter(last_nha_yrs > dx_crpc_yrs) %>%
  filter(last_nha_yrs > dx_cpt_best_yrs) %>%
  nrow(.)

n_nha_post_mcrpc_seq_60d <- dft_joined_times %>%
  filter(!is.na(dx_dmet_yrs) & !is.na(dx_crpc_yrs) & !is.na(first_nha_yrs)) %>%
  filter(last_nha_yrs > dx_dmet_yrs) %>%
  filter(last_nha_yrs > dx_crpc_yrs) %>%
  filter(last_nha_yrs + (60 / 365.25) > dx_cpt_best_yrs ) %>%
  nrow(.)
  
```

## Project 1

Analysis description from Dec 1, 2023 email (see below fore second project):

> 1) Interrogating how the prognostic value of BRCA2mut is modulated by other events:
>
> - Patients who have received a novel hormonal agent (NHA: including abiraterone, enzalutamide, darolutamide, apalutamide) in the mCRPC setting 
> - In those, we want to interrogate “composite” biomarkers based on BRCA2, TP53 and RB1 genomic status.
> - Clinical outcome data we would like to correlate the composite biomarkers with: response (PSA50%), PFS to the NHA, overall survival.

### Cohort size   

The definition of an NHA (novel hormonal agent) is taken directly from the drugs listed in the email with small alterations to match the genie encoding (e.g. Abiraterone acetate" instead of "abiraterone").  Here are the relevant numbers of participants:

- There are `r n_cohort` people in the BPC prostate dataset.
- Of those, `r n_mcrpc` were diagnosed both a distant metastasis and castrate-resistant prostate cancer (via medical oncologist notes).  This includes de novo metastatic cases.
- **Among mCRPC cases, `r n_nha_post_mcrpc` had exposure to a NHA in the mCRPC period**
  - *Aside:* Among mCRPC cases, `r n_nha_anytime_mcrpc` had exposure to NHA at any time.
  
The next part of the question deals with the timing of the NGS (next generation sequencing), and there are multiple dates recorded in GENIE BPC.  We use the order date if it is available.  For about 5% of the NGS tests in our cohort (especially UHN, in Canada) the order date is not available and we use the date of the NGS report.  The difference between these dates is typically about 40 days.  Here are the numbers bringing this data in:

- **Among those with NHA exposure in the mCRPC period, `r n_nha_post_mcrpc_seq` were sequenced before the last day of exposure to NHA.**
  - *Variant*: `r n_nha_post_mcrpc_seq_60d` were sequenced within 60 days of the last day of exposure to NHA.
  
### Notes

- We noted in our work that use of CRPC flags differs greatly by institution.  Some declare it quickly and some do not.  This may be worth examining in greater detail to assess whether this is a helpful criterion on top of metastasis.  Some patients have long periods of followup where neither hormone-sensitive or castrate-resistant are marked in our datasets.
- We used the "lastadm" versions of the drug variables, which encode either the discontinuation date or the last known administration depending on what is available.  This variable is more complete than relying on firmly declared discontinuation dates.
- The vast majority of the NGS panels used in this cohort cover BRCA2, TP53 and RB1.The exception is the small panels from UHN (UHN-50-V2 and UHN-48-V1), which are missing BRCA2.  These cover about 30 samples together, so the impact should be minor.  Imputation or dropping incompletely tested cases are obviously both options depending on your preference.
- We have not evaluated how the timing of the **first** NHA use relates to anything.  This only answers whether the period(s) of NHA use have any overlap with other intervals.  For example, it might be the third NHA use for a patient that overlaps with mCRPC, which I'm pointing out for the clinical relevance.

## Project 2

Analysis description from Dec 1, 2023 email (second project only):

> 2) Genomic landscape of the post-NHA prostate cancer
> - Number of patients who received a NHA in the hormone-naive setting and had a biopsy after developing CRPC
> - Number of patients who did not receive an NHA in the hormone-naive setting and had a biopsy after developing CRPC (but before receiving an NHA in the CRPC setting)
> - Number of patients who received a NHA in the castration-resistant setting and had a biopsy after developing resistance to NHA
> 
> As said, this project 2 is probably too challenging as few patients are re-biopsied post-NHA and also because NHA only were reimbursed for hormone- naive patients from 2018ish onwards, and the PFS is around 3 years, so you may have not that many of them in the study.


### Clarifications

A clarification is needed on hormone-naive:  *How should we define the hormone-naive setting?*  The two concepts we have in PRISSMM during medical oncologist assessments are:

- **HSPC** is equivalent to hormone sensitive = castrate sensitive = androgen sensitive prostate cancer.
- **CRPC** is equivalent to hormone-resistant = castrate-resistant = androgen-resistant = independent prostate cancer.

Today I am proceeding under the assumption that hormone-naive and hormone-sensitive are equivalent.  See notes for a few other assumptions.  

As far as I know, we do not have documentation of resistance to NHA, so I will skip the third question for now.  Big assumptions (for example, that any discontinuation is due to resistance) may make this feasible if those are acceptable for your aims.


```{r}
dft_crpc_block <- make_cast_status_block(
  dft_med_onc,
  dft_ca_ind
) %>%
  select(-contains("os_dx")) %>%
  rename(
    dx_block_start_yrs = dx_block_start,
    dx_block_end_yrs = dx_block_end
  )

dft_crpc_block_miss_excl <- dft_crpc_block %>% 
  filter(!(md_cast_status_f %in% "Not yet declared"))

dft_biopsy_t <- dft_path %>%
  filter(path_proc %in% "Biopsy") %>%
  select(record_id, path_proc_int)

dft_biopsy_t <- dft_ca_ind %>% 
  select(record_id, dob_ca_dx_yrs) %>%
  distinct(.) %>%
  left_join(
    dft_biopsy_t,
    ., 
    by = "record_id"
  ) 

dft_biopsy_t %<>%
  mutate(
    dx_path_proc_yrs = (path_proc_int / 365.25) - dob_ca_dx_yrs
  ) %>%
  select(-path_proc_int, -dob_ca_dx_yrs)

```


```{r}
eval_event_in_interval <- function(
    dat_block,
    dat_event,
    var_event_t,
    var_block_group = 'md_cast_status_f',
    tol = 0.5 / 365.25, # half a day in years tolerance
    return_type = "summary"
) {
  
  dat_rtn <- left_join(
    dat_block,
    select(dat_event, record_id, all_of(var_event_t)),
    by = c("record_id"),
    # up to two blocks can match infinitely many events per person
    relationship = "many-to-many"
  )
  
  dat_rtn %<>%
    group_by(record_id, .data[[var_block_group]]) %>%
    mutate(
      lt_end = dx_block_end_yrs > (.data[[var_event_t]] - tol),
      gt_start = dx_block_start_yrs < (.data[[var_event_t]] + tol),
      event_in_block = lt_end & gt_start
    ) %>%
    ungroup(.) %>%
    select(-lt_end, -gt_start)
  
  
  if (return_type %in% "full") {
    return(dat_rtn)
  }
  else if (return_type %in% "summary") {
    dat_rtn %<>%
      group_by(record_id, .data[[var_block_group]]) %>%
      summarize(
        num_events_in_block = sum(event_in_block),
        any_events_in_block = num_events_in_block >= 1,
        .groups = 'drop'
      )
    return(dat_rtn)
  } else {
    cli::cli_abort("Invalid return_type")
  }
  
}

dft_biopsy_in_block <- eval_event_in_interval(
  dat_block = dft_crpc_block_miss_excl,
  dat_event = dft_biopsy_t,
  var_event_t = "dx_path_proc_yrs"
) 

dft_biopsy_in_block_full <- eval_event_in_interval(
  dat_block = dft_crpc_block_miss_excl,
  dat_event = dft_biopsy_t,
  var_event_t = "dx_path_proc_yrs",
  return_type = "full"
) 
```


```{r}
eval_drug_in_interval <- function(
    dat_block,
    dat_drug,
    var_block_group = "md_cast_status_f",
    tol = 0.5 / 365.25
) {
  
  dat_drug %<>%
    mutate(
      dx_drug_start_int_yrs = dx_drug_start_int / 365.25,
      dx_drug_end_or_lastadm_int_yrs = dx_drug_end_or_lastadm_int / 365.25
    ) %>%
    select(
      record_id,
      dx_drug_start_int_yrs,
      dx_drug_end_or_lastadm_int_yrs
    )
  
  dat_rtn <- left_join(
    dat_block,
    dat_drug,
    # we can get away with just record ID due to the cohort (no double index cancers)
    by = c("record_id"),
    # up to two blocks match infinitely many drug rows:
    relationship = "many-to-many" 
  )
  
  dat_rtn %<>%
    group_by(record_id, .data[[var_block_group]]) %>%
    mutate(
      incomplete_drug_data = is.na(dx_drug_start_int_yrs) | is.na(dx_drug_end_or_lastadm_int_yrs),
      overlap_cond_1 = (dx_block_end_yrs - dx_drug_start_int_yrs > tol) &
        (dx_block_start_yrs - dx_drug_start_int_yrs < tol),
      overlap_cond_2 = (dx_block_start_yrs - dx_drug_end_or_lastadm_int_yrs < tol) &
        (dx_block_start_yrs - dx_drug_start_int_yrs > tol),
      overlap = case_when(
        incomplete_drug_data ~ F,
        overlap_cond_1 | overlap_cond_2 ~ T,
        T ~ F # no known cases
      )
    )
  
  dat_rtn %<>%
    group_by(record_id, .data[[var_block_group]]) %>%
    summarize(
      num_overlap = sum(overlap, na.rm = T),
      any_overlap = num_overlap >= 1,
      .groups = "drop"
    )
  
  return(dat_rtn)
  
}    

dft_nha_in_block <- eval_drug_in_interval(
  dat_block = dft_crpc_block_miss_excl,
  dat_drug = dft_drug_nha
) 

dft_block_info <- full_join(
  # renaming here to avoid confusion:
  (dft_nha_in_block %>%
     select(record_id, md_cast_status_f,
            any_drug_overlap = any_overlap)
  ),
  (dft_biopsy_in_block %>%
     select(record_id, md_cast_status_f,
            any_biopsy_in_block = any_events_in_block)
  ),
  by = c('record_id', 'md_cast_status_f'),
  relationship = 'one-to-one'
)

vec_nha_in_hspc <- dft_block_info %>%
  filter(md_cast_status_f %in% "Hormone Sensitive") %>%
  filter(any_drug_overlap) %>%
  pull(record_id) %>%
  unique

vec_biopsy_in_crpc <- dft_block_info %>%
  filter(md_cast_status_f %in% "Castrate-Resistant") %>%
  filter(any_biopsy_in_block) %>%
  pull(record_id) %>%
  unique


```

### Cohort size

#### First question

- Number of patients who had NHA in the HSPC setting: `r length(vec_nha_in_hspc)`
- Number of patients who had a biopsy in the CRPC setting: `r length(vec_biopsy_in_crpc)`
- Number who meet both criteria: `r length(intersect(vec_nha_in_hspc, vec_biopsy_in_crpc))`

```{r}
dft_first_crpc_biopsy <- dft_biopsy_in_block_full %>%
  filter(md_cast_status_f %in% "Castrate-Resistant") %>%
  filter(event_in_block) %>%
  group_by(record_id) %>%
  arrange(dx_path_proc_yrs) %>%
  slice(1) %>%
  ungroup(.)

dft_project_2_2 <- dft_drug_nha_first_last %>%
  select(record_id, first_nha_yrs) %>%
  left_join(
    dft_first_crpc_biopsy,
    .,
    by = "record_id"
  )

n_project_2_2 <- dft_project_2_2 %>% 
  filter(first_nha_yrs > dx_path_proc_yrs | is.na(first_nha_yrs)) %>%
  nrow(.)

```


#### Second question  

I'm going to rephrase this slightly: "Number of patients who had a biopsy after developing CRPC and had no receipt of NHA before this biopsy date".  The difference here is mostly how we count the periods where neither HSPC nor CRPC have been declared.  If this alteration is not acceptable just let me know.

- Number who had a biopsy in the CRPC setting: `r nrow(dft_project_2_2)`
- Number who had no NHA before that biopsy:  `r n_project_2_2`


### Notes 

- **Missing CRPC/HSPC (high impact)** When I say "HSPC" or "CRPC" I am excluding any period of followup where the medical oncologist did not yet declare HSPC or CRPC.  Excluding undeclared time blocks has a major impact because that is a common situation in our cohort.  An alternative assumption might be assuming that no note means HSPC.
- Once someone is declared as having CRPC I assume this condition is permanent.  For example, if the next assessment is blank or differs, I ignore that and consider them castrate-resistant still.
