---
title: "Feasibility report"
subtitle:  "Joaquin Mateo analyses"
author: "Alex Paynter"
date: "`r Sys.Date()`"
output: 
  html_document:
    toc: true
    toc_float: true
    theme: paper
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = F, include = F,
                      warning = F, message = F,
                      fig.width = 7, fig.height = 5)
k_dpi <- 150
pal_pt_hc <- c("#004488","#ddaa33", "#bb5566")
```

<style type="text/css">
.main-container {
  max-width: 1000px !important;
  margin: auto;
}
</style>

```{r}
library(purrr)
library(here)
library(fs)

purrr::walk(.x = fs::dir_ls(here('R')), .f = source) # also loads lots of packages.
```



```{r, set_gtsummary_theme}
theme_gtsummary_compact()
theme_gtsummary_language("en", big.mark = "") # for now.


```


```{r, warning = T}
read_wrap_clin <- function(p) {
  read_rds(file = here("data", 'clin', p))
}

dft_pt <- read_wrap_clin("dft_pt.rds")
dft_ca_ind <- read_wrap_clin("dft_ca_ind.rds")
# we're taking the augmented version, which has TMB columns added.  The existing info is the same.
dft_cpt <- read_wrap_clin("dft_cpt_aug.rds")
dft_reg <- read_wrap_clin("dft_reg.rds")
dft_med_onc <- read_wrap_clin('dft_med_onc.rds')
dft_drug <- read_csv(
  file = here('data', 'drug.csv')
)
```


```{r, warning = F}
dft_drug %<>% 
  mutate(
    agent = str_replace_all(drug, "\\(.*", "")
  ) %>%
  relocate(
    agent,
    .before = drug
  )

vec_agent_nha <- c(
  'Abiraterone Acetate',
  'Enzalutamide',
  'Darolutamide',
  'Apalutamide'
)

dft_drug_nha <- dft_drug %>%
  filter(agent %in% vec_agent_nha)

dft_drug_nha_first_last <- dft_drug_nha %>% 
  group_by(record_id, ca_seq) %>%
  summarize(
    nha_used = paste(sort(unique(agent)), collapse = ", "),
    first_nha_yrs = min(dx_drug_start_int, na.rm = T) / 365.25,
    last_nha_yrs = max(dx_drug_end_or_lastadm_int, na.rm = T) / 365.25,
    .groups = 'drop'
  )
```


```{r}
dft_dmet_time <- make_dmet_status_block(ca_ind_dat = dft_ca_ind) %>%
  filter(dmet_status %in% 'Distant Metastasis') %>%
  rename(dx_dmet_yrs = dx_block_start) %>%
  select(record_id, ca_seq, dx_dmet_yrs)

dft_crpc_time <- make_cast_status_block(
  dft_med_onc,
  dft_ca_ind
) %>%
  filter(md_cast_status_f %in% 'Castrate-Resistant') %>%
  rename(dx_crpc_yrs = dx_block_start) %>%
  select(record_id, dx_crpc_yrs)

dft_first_cpt <- full_join(
  get_first_cpt(dft_ca_ind, dft_cpt, type = 'order'),
  get_first_cpt(dft_ca_ind, dft_cpt, type = 'report'),
  by = c('record_id', 'ca_seq')
) %>%
  mutate(
    dx_cpt_best_yrs = if_else(
      is.na(dx_cpt_order_yrs),
      dx_cpt_rep_yrs,
      dx_cpt_order_yrs
    )
  )
  
  

dft_joined_times <- full_join(
  dft_drug_nha_first_last,
  dft_dmet_time,
  by = c("record_id", "ca_seq"),
  relationship = 'one-to-one'
) %>%
  full_join(
    dft_first_cpt,
    by = c("record_id", 'ca_seq'),
    relationship = 'one-to-one'
  ) %>%
  full_join(
    dft_crpc_time,
    by = 'record_id',
    relationship = 'one-to-one'
  )

n_nha <- dft_joined_times %>%
  filter(!is.na(first_nha_yrs)) %>%
  nrow(.)

n_mcrpc <- dft_joined_times %>%
  filter(!is.na(dx_dmet_yrs) & !is.na(dx_crpc_yrs)) %>%
  nrow(.)

n_nha_anytime_mcrpc <- dft_joined_times %>%
  filter(!is.na(dx_dmet_yrs) & !is.na(dx_crpc_yrs) & !is.na(first_nha_yrs)) %>%
  nrow(.)

n_nha_post_mcrpc <- dft_joined_times %>%
  filter(!is.na(dx_dmet_yrs) & !is.na(dx_crpc_yrs) & !is.na(first_nha_yrs)) %>%
  filter(last_nha_yrs > dx_dmet_yrs) %>%
  filter(last_nha_yrs > dx_crpc_yrs) %>%
  nrow(.)

n_nha_post_mcrpc_seq <- dft_joined_times %>%
  filter(!is.na(dx_dmet_yrs) & !is.na(dx_crpc_yrs) & !is.na(first_nha_yrs)) %>%
  filter(last_nha_yrs > dx_dmet_yrs) %>%
  filter(last_nha_yrs > dx_crpc_yrs) %>%
  filter(last_nha_yrs > dx_cpt_best_yrs) %>%
  nrow(.)

n_nha_post_mcrpc_seq_60d <- dft_joined_times %>%
  filter(!is.na(dx_dmet_yrs) & !is.na(dx_crpc_yrs) & !is.na(first_nha_yrs)) %>%
  filter(last_nha_yrs > dx_dmet_yrs) %>%
  filter(last_nha_yrs > dx_crpc_yrs) %>%
  filter(last_nha_yrs + (60 / 365.25) > dx_cpt_best_yrs ) %>%
  nrow(.)
  
```

## Analysis 1

Analysis description from Dec 1, 2023 email:

> 1) Interrogating how the prognostic value of BRCA2mut is modulated by other events:
>
> - Patients who have received a novel hormonal agent (NHA: including abiraterone, enzalutamide, darolutamide, apalutamide) in the mCRPC setting - In those, we want to interrogate “composite” biomarkers based on BRCA2, TP53 and RB1 genomic status.
> - Clinical outcome data we would like to correlate the composite biomarkers with: response (PSA50%), PFS to the NHA, overall survival.

### Cohort size   

- There are `r n_cohort` people in the BPC prostate dataset.
- Of those, `r n_mcrpc` were diagnosed both a distant metastasis and castrate-resistant prostate cancer (via medical oncologist notes).  This includes de novo metastatic cases.
- **Among mCRPC cases, `r n_nha_post_mcrpc` had exposure to a NHA in the mCRPC period**
  - Among mCRPC cases, `r n_nha_anytime_mcrpc` had exposure to NHA at any time.
  
To bring in the timing of next generation sequencing we use the order date if it is available.  For about 5% of the NGS tests in our cohort (especially UHN, in Canada) the order date is not available and we use the date of the NGS report.  The difference between these dates is typically about 40 days.  Here are the numbers bringing this data in:

- **Among those with NHA exposure in the mCRPC period, `r n_nha_post_mcrpc_seq` were sequenced before the last day of exposure to NHA.**
  - Variant: `n_nha_post_mcrpc_seq_60d` were sequenced within 60 days of the last day of exposure to NHA.
  
Notes:

- We noted in our work that use of CRPC flags differs greatly by institution.  Some declare it quickly and some do not.  This may be worth examining in greater detail to assess whether this is a helpful criterion on top of metastasis.  Some patients have long periods of followup where neither hormone-sensitive or castrate-resistant are marked in our datasets.
- We used the "lastadm" versions of the drug variables, which encode either the discontinuation date or the last known administration depending on what is available.  This variable is more complete than relying on firmly declared discontinuation dates.
- The vast majority of the panels used in this cohort cover BRCA2, TP53 and RB1.The exception is the small panels from UHN (UHN-50-V2 and UHN-48-V1), which are missing BRCA2.  These cover about 30 samples together, so the impact should be minor, but I have not specifically removed these samples in the analysis above.  
- We have not evaluated how the timing of the **first** NHA use relates to anything.  This only answers whether the period(s) of NHA use have any overlap with other intervals.  For example, it might be the third NHA use for a patient that overlaps with mCRPC, which I'm pointing out for the clinical relevance.
