---
title: "GENIE BPC Prostate Cancer - Landscape Report"
author: "Alex Paynter"
date: "Started April 4, 2023"
output: 
  html_document:
    toc: true
    toc_float: true
    theme: paper
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = F, include = F,
                      warning = F, message = F,
                      fig.width = 7, fig.height = 5)
k_dpi <- 150
```


<style type="text/css">
.main-container {
  max-width: 1000px !important;
  margin: auto;
}
</style>



```{r}
library(cli)
library(readr)
library(here)
library(dplyr)
library(tidyr)
library(purrr)
library(rlang)
library(magrittr)
library(janitor)
library(stringr)
library(vctrs)
library(glue)

library(ggplot2)
library(ggrepel)
library(ggtext)

library(gt)
library(gtsummary)
library(sunburstR)
library(huxtable)


as_list(here("R", dir(here("R"))))

# Source everything in the "R" folder - just functions, no scripts.
purrr::walk(.x = here("R", dir(here("R"))), .f = source)
```


```{r, set_gtsummary_theme}
theme_gtsummary_compact()
theme_gtsummary_language("en", big.mark = "") # for now.
```


```{r, warning = T}
read_wrap <- function(p) {
  read_csv(file = here("data-raw", p), show_col_types = F)
}

dft_pt <- read_wrap("patient_level_dataset.csv")
dft_ca_ind <- read_wrap("cancer_level_dataset_index.csv")
dft_img <- read_wrap("imaging_level_dataset.csv")
dft_med_onc <- read_wrap("med_onc_note_level_dataset.csv")
dft_path <- read_wrap("pathology_report_level_dataset.csv")
dft_regimens <- read_wrap("regimen_cancer_level_dataset.csv")
dft_tm <- read_wrap("tm_level_dataset.csv")
dft_cpt <- read_wrap("cancer_panel_test_level_dataset.csv")

# A few sanity checks on the data:
if ((dft_pt$record_id %>% duplicated %>% any)) {
  stop("Duplicated records in patient level dataset.")
}

if (any(dft_pt$naaccr_sex_code != "Male")) {
  warning("Non-male sex detected for at least one participant (unexpectedly for Prostate cancer)")
}

```


```{r, demo_data_manipulation, include = F}
dft_pt_baseline_sub <- dft_pt %>%
  mutate(
    `Race (primary)` = format_ptlevel_naaccr_race_code_primary(
      naaccr_race_code_primary
    ),
    `Ethnicity` = format_ptlevel_naaccr_ethnicity_code(
      naaccr_ethnicity_code
    )) %>%
  select(
    record_id, 
    Institution = institution,
    `Race (primary)`,
    `Ethnicity`,
    birth_year
  )


dft_ca_ind_baseline_sub <- dft_ca_ind %>%
  mutate(
    ca_dx_how = format_ca_dx_how(ca_dx_how),
    ca_hist_adeno_squamous = format_ca_hist_adeno_squamous(ca_hist_adeno_squamous),
    stage_dx = format_stage_dx(stage_dx),
    ca_path_n_stage = format_ca_path_n_stage(ca_path_n_stage)
  ) %>%
  select(
    record_id,
    `Age at dx (years)` = dob_ca_dx_yrs,
    `Stage at dx` = stage_dx,
    `Source of dx` = ca_dx_how,
    # Update May 5:  Removed histology (unimportant)
    # `Histology` = ca_hist_adeno_squamous, # at dx?
    `Pathologic N Stage` = ca_path_n_stage
  ) 

# On histology alternatives:
# - naaccr_histology_cd has mostly the same info according to https://seer.cancer.gov/icd-o-3/.  Most of these levels are singletons.



dft_tm_baseline <- dft_tm %>%
  psa_test_unit_convert() %>%
  filter(!is.na(tm_num_result_conv)) %>%
  # filter down to results at or on diagnosis.
  # Importantly there are psa/test values available before diagnosis for
  #   many participants.
  filter(dx_tm_days <= 30 & dx_tm_days >= -180) %>% 
  group_by(record_id, tm_type) %>%
  # Prioritize values at/before 
  mutate(abs_dx_tm_days = abs(dx_tm_days)) %>%
  arrange(abs_dx_tm_days) %>%
  slice(1) %>%
  ungroup() %>% 
  select(record_id, 
         tm_type, 
         tm_num_result_conv, 
         tm_result_units_conv) 

# Visual check:
# ggplot(dft_tm_baseline, aes(x = dx_tm_days)) + stat_ecdf()


# kind of a hacky check:  the resulting tibble will be 2 rows and 3 columns if there are 2 analytes with one unit each.
tm_unit_conv_check <- dft_tm_baseline %>% 
  tabyl(tm_type, tm_result_units_conv) %>% 
  dim
if (!all.equal(tm_unit_conv_check,c(2,3))) {
  cli::abort("Unit conversion problem.")
}

dft_tm_baseline %<>% 
  select(-tm_result_units_conv) %>%
  tidyr::pivot_wider(
    names_from = "tm_type",
    values_from = "tm_num_result_conv"
  ) %>%
  # Not needed, decision made 2023 May 8
  select(-Testosterone) %>% 
  rename(
    `PSA (ng/mL)` = PSA,
    #`Testosterone (ng/dL)` = Testosterone
  )










dft_demo <- full_join(
  dft_pt_baseline_sub,
  dft_ca_ind_baseline_sub,
  by = "record_id"
) %>%
  full_join(
    .,
    dft_tm_baseline,
    by = "record_id"
  )

# age_dx is not an integer in this cohort, so this should be more exact.
dft_demo %<>% 
  mutate(
    `Year of birth` = birth_year,
    `Year of diagnosis` = round(birth_year + `Age at dx (years)`)
  ) %>%
  select(-birth_year) %>%
  relocate(`Year of birth`, `Year of diagnosis`,
           .before = `Age at dx (years)`) 

```



## Content

### *R1.1a: Characteristics at dx

For most analyses, especially survival, the timing of observations is critical.  Here we have separated characteristics at diagnosis (Table R1.1), characteristics at sequencing, and characteristics which could represent a first post-baseline observation (Table R1.3). 

Notes:

- For PSA, an arbitrary choice was made to limit observations to those from 180 days before to 30 days after diagnosis of prostate cancer.  Notably this leaves the majority of the cohort unobserved at baseline on these variables.  See appendix for unit conversion assumptions.
- Year of diagnosis is approximate.  It is estimated from birth year and interval to diagnosis, so it could be off by one year in either direction for any particular subject.  All years in the report suffer from this problem.



```{r, tab_r1_1a, include = T}
dft_demo %>%
  select(-record_id) %>%
  gtsummary::tbl_summary(
    data = .,
    digits = list(
      `Year of birth` ~ 0,
      `Year of diagnosis` ~ 0
    )
  )
```

### *R1.1b: Char. at dx by site

Observations:

- MSK, UHN and VICC have more late stage cancers (about 50%).  DFCI has more Stage II/III cancers.
- Several variables are 100% unknown at UHN:  Histology and Ethnicity.

```{r, tab_r1_1b, include = T}
dft_demo %>%
  select(-record_id) %>%
  gtsummary::tbl_summary(
    data = ., 
    by = Institution,
    digits = list(
      `Year of birth` ~ 0,
      `Year of diagnosis` ~ 0
    )
  )
```












```{r, create_seq_data_table}
# Table of characteristics at sequencing.
dft_char_seq <- get_first_cpt(dft_ca_ind, dft_cpt)

# Figure out if the participants were castrate resistant at seq time:
dft_cast_block_seq <- make_cast_status_block(
  dft_med_onc, 
  dft_ca_ind, 
  remove_hspc_after_crpc = T) %>%
  select(record_id, md_cast_status_f, matches("^dx_block"))
dft_cast_block_seq %<>% left_join(
  .,
  select(dft_char_seq, record_id, dx_cpt_rep_yrs),
  by = "record_id",
  relationship = "many-to-one"
) %>%
  filter(dx_cpt_rep_yrs >= dx_block_start & 
           dx_cpt_rep_yrs < dx_block_end)
# count(dft_char_seq, record_id, ca_seq, sort = T)
dft_cast_block_seq %<>% 
  select(-matches("^dx_block"), -dx_cpt_rep_yrs)
dft_char_seq %<>% left_join(., dft_cast_block_seq, by = "record_id")
dft_char_seq %<>%
  mutate(md_cast_status_f = forcats::fct_na_value_to_level(
    md_cast_status_f,
    level = "Not yet declared")
  )


# Figure out if distant metastasis has been diagnosed by sequencing:
dft_char_seq <- make_dmet_status_block(ca_ind_dat = dft_ca_ind) %>% 
  filter(dmet_status %in% "Distant Metastasis") %>%
  select(record_id, ca_seq, dmet_start = dx_block_start) %>%
  left_join(
    dft_char_seq,
    .,
    by = c("record_id", "ca_seq"),
    relationship = "one-to-one"
  )
dft_char_seq %<>% 
  mutate(dmet_at_seq = case_when(
    is.na(dmet_start) ~ "No",
    dmet_start > dx_cpt_rep_yrs ~ "No",
    T ~ "Yes"
  )) %>% 
  select(-dmet_start)

# Pull in some additional variables from other datasets for time calcs:
dft_char_seq <- dft_pt %>%
  select(record_id, birth_year, institution) %>%
  left_join(
    dft_char_seq,
    .,
    by = "record_id"
  )
dft_char_seq <- left_join(
  dft_char_seq,
  select(
    dft_ca_ind, record_id, ca_seq, dob_ca_dx_yrs
  ),
  by = c("record_id", "ca_seq")
)
dft_char_seq %<>%
  mutate(
    seq_year = round(birth_year + dob_ca_dx_yrs + dx_cpt_rep_yrs),
  ) %>% 
  select(
    record_id,
    Institution = institution,
    `Year of sequencing` = seq_year,
    `Dx to seq. (years)` = dx_cpt_rep_yrs,
    `Distant metastasis at seq.` = dmet_at_seq,
    `Castration status at seq.` = md_cast_status_f
  )
    

```

### *R1.2a: Characteristics at sequencing

The following table displays cohort characteristics at the time of sequencing.  The year of sequencing is approximate (could be off by one year in either direction).

```{r, output_char_at_seq, include = T}
dft_char_seq %>%
  select(-c(record_id, Institution)) %>%
  gtsummary::tbl_summary(
    data = .,
    digits = list(
      `Year of sequencing` ~ 0
    )
  )
```

### *R1.2b: Char. at seq. by site

The following table displays cohort characteristics at the time of sequencing.  The year of sequencing is approximate (could be off by one year in either direction).

```{r, output_char_at_seq_by_site, include = T}
dft_char_seq %>%
  select(-record_id) %>%
  gtsummary::tbl_summary(
    data = .,
    digits = list(
      `Year of sequencing` ~ 0
    ),
    by = "Institution"
  )
```

**Observation:** DFCI has more early stage patients, and a lower median time to sequencing.  This gives me the impression that they may tend to sequence patients early in disease progression.





```{r, create_post_dx_char}
dft_ca_ind_post_baseline <- dft_ca_ind %>%
  mutate(
    ce_num = format_cores(ca_pros_num_cores),
    cp_num = format_cores(ca_pros_num_corespos),
    clin_gleason = format_gleason(ca_pros_clin_gscore)
  ) %>%
  select(record_id, ce_num, cp_num, clin_gleason)



dft_med_onc_post_baseline <- dft_med_onc %>%
  group_by(record_id) %>%
  arrange(md_visit_number) %>%
  # filter down to the first determination made.
  #  That is, we will ignore rows with "Not stated in the Impression/Plan", unless that is the only type of entry ever
  #  Made for the patient.
  filter(str_detect(md_pca_status, "(CRPC)|(HSPC)") | 1:n() == n()) %>%
  slice(1) %>%
  ungroup %>%
  # First castration-resistance determination:
  mutate(first_cast_det = if_else(str_detect(md_pca_status, "Not stated in"),
                              "Never stated in Impression/Plan",
                              md_pca_status),
         first_cast_det = factor(first_cast_det)) %>%
  select(record_id,
         first_cast_det)

# Dominant units for PSA are ng/mL
# Dominant units for Testosterone are ng/dL

tm_missing_units <- dft_tm %>%
  filter(tm_type %in% c("PSA", "Testosterone")) %>%
  group_by(tm_type) %>%
  filter(is.na(tm_result_units) & !is.na(tm_num_result)) %>%
  summarize(n = n())

# # Visual check on the unit conversions if desired.  Just looking for
# good overall overlap between converted and non-converted units.
# dft_tm %>%
#   psa_test_unit_convert() %>%
#   mutate(converted_unit = tm_result_units != tm_result_units_conv) %>%
#   ggplot(., aes(x = 1, y = tm_num_result_conv, color = converted_unit)) + 
#   geom_jitter(width = 1, height = 0) + 
#   scale_y_log10() + 
#   facet_wrap(vars(tm_type))



# dft_ca_ind_post_baseline %>% count(record_id) %>% arrange(desc(n))
# dft_med_onc_post_baseline %>% count(record_id) %>% arrange(desc(n))
# dft_ca_ind_post_baseline %>% count(record_id) %>% arrange(desc(n))


dft_post_baseline <- full_join(
  dft_ca_ind_post_baseline,
  dft_med_onc_post_baseline,
  by = "record_id"
) 

dft_post_baseline %<>%
  left_join(
    ., 
    select(dft_pt, record_id, institution),
    by = "record_id"
  )

dft_post_baseline %<>%
  mutate(
    first_cast_det = forcats::fct_na_value_to_level(
      f = first_cast_det, 
      level = "Zero med onc notes"
    )
  )

dfp_post_baseline <- dft_post_baseline %>%
  select(
    record_id,
    Institution = institution,
    # Update: May 5 - remove, not needed.
    # `Prostate cores examined` = ce_num,
    # `Prostate cores positive` = cp_num,
    `Clinical Gleason score` = clin_gleason,
    `Castration status (first)` = first_cast_det,
  )

dfp_post_baseline %<>%
  mutate()
  

```


### R1.3a: Characteristics (dx onward)

The characteristics in this table were known to be curated from diagnosis to the end of surgeries in the first round of treatment.  In the case of castration status, we take the first value where CRPC/HRPC was declared (that is, ignore the records where no determination was made).


```{r, tab_r1_2a, include = T}
dfp_post_baseline %>%
  select(-c(record_id, Institution)) %>%
  gtsummary::tbl_summary(data = .)
```

### R1.3b: Characteristics (dx onward) by site

Observations on the table:

- DFCI has more participants with low first observed Gleason scores.  This is consistent with the early stage observation from the previous table.
- MSK has more participants with HSPC on the first observation, which seems inconsistent with higher average stage.  This may indicate a scoring difference on HSPC/CRPC between institutions.

```{r, tab_r1_2b, include = T}
dfp_post_baseline %>%
  select(-record_id) %>%
  gtsummary::tbl_summary(data = ., by = Institution)
```







```{r}
# Just checking an assumption I have about the data:
if (nrow(
  (dft_med_onc %>% filter(!(record_id %in% dft_ca_ind$record_id)))
) != 0) {
  cli::cli_abort("Med onc data contains partipants not in the index cancer dataset.")
}

dft_count_all <- make_obs_count_df(
  med_onc_dat = dft_med_onc,
  img_dat = dft_img,
  reg_dat = dft_regimens,
  ca_ind_dat = dft_ca_ind
)

obs_type_levs <- c("Med Onc Notes",
                   "Regimens",
                   "Total Scans",
                   "CT Scan",
                   "MRI Scan",
                   "PET or PET-CT")
                   
dft_count_long <- dft_count_all %>%
  select(record_id, 
         n_med_onc, 
         n_scan_ct, 
         n_scan_mri, 
         n_scan_pet_or_pet_ct,
         n_scan_total,
         n_regimens) %>%
  pivot_longer(cols = -record_id,
               names_to = "type",
               values_to = "n") %>%
  mutate(
    type_f = case_when(
      type %in% "n_med_onc" ~ obs_type_levs[1],
      type %in% "n_regimens" ~ obs_type_levs[2],
      type %in% "n_scan_total" ~ obs_type_levs[3],
      type %in% "n_scan_ct" ~ obs_type_levs[4],
      type %in% "n_scan_mri" ~ obs_type_levs[5],
      type %in% "n_scan_pet_or_pet_ct" ~ obs_type_levs[6]
    ),
    type_f = factor(type_f, levels = obs_type_levs)
  )

dft_count_quantiles <- dft_count_long %>%
  nest(.by = type_f) %>%
  mutate(res = purrr::map(.f = get_quantile_df,
                          .x = data,
                          var = "n")
  ) %>%
  select(type_f, res) %>%
  unnest(res) 

# Really like this plot but nobody else in my group will:
#
# gg_count_ecdf_2 <- ggplot(data = dft_count_long,
#          aes(x = n)) + 
#   stat_ecdf(aes(color = type_f)) + 
#   geom_label_repel(data = dft_count_quantiles,
#              # yes, this is screwed up:
#              aes(x = y, y = n, label = y, color = type_f),
#              hjust = 0, nudge_x = 50, size = 2.5,
#              direction = "x",
#              segment.size = 0.15,
#              show_guide = F
#             ) + 
#   theme_bw() +
# #  facet_wrap(vars(type_f), scales = "free", nrow = 2) + 
#   scale_y_continuous(
#     name = "Cohort fraction",
#     breaks = c(0, .5, 1),
#     labels = paste0(c(0, .5, 1)*100, "%")
#   ) + 
#   theme(
#     strip.text = element_text(hjust = 0),
#     legend.position = "bottom"
#   ) +
#   guides(
#     color = guide_legend(
#       title = NULL,
#       override.aes = aes(label = "", size = 5, shape = 15)
#     ),
#   ) + 
#   scale_color_manual(
#     values = c('#EE7733', '#0077BB', '#33BBEE', '#EE3377', '#CC3311', '#009988', '#BBBBBB')
#   )
# 
# gg_count_ecdf_2

 gg_count_ecdf <- ggplot(data = dft_count_long,
         aes(x = n)) + 
  stat_ecdf(color = '#4477AA') + 
  geom_text_repel(data = dft_count_quantiles,
             # yes, this is screwed up:
             aes(x = y, y = n, label = y),
             hjust = 0, nudge_x = 8, size = 2.5,
             direction = "x",
             segment.size = 0.15
            ) + 
  theme_bw() +
  facet_wrap(vars(type_f), scales = "free", nrow = 2) + 
  scale_y_continuous(
    name = "Cohort fraction",
    breaks = c(0, .5, 1),
    labels = paste0(c(0, .5, 1)*100, "%")
  ) + 
  theme(
    strip.text = element_text(hjust =0)
  )


gg_count_hist_quan <- plot_count_hist_quan(
  dat = dft_count_long,
)
  
  
```


### R1.4a Observation counts (eCDF)

The follow plot shows the empirical cumulative distirbution function for the number of observations per subject.  In each plot, three numbers are labelled:  the 25th percentile, the median and the 75th percentile.  

In addition to the basic stats this gives our intended audience for this plot (future BPC data users) information they may need to know in fitting their models.  For example, you can tell how many people have at least 3 regimens, or roughly how many people have over 75 medical oncology notes.

```{r, output_obs_ecdf, include = T}
gg_count_ecdf
```


### R1.4b Observation counts (histogram)

```{r, include = T}
gg_count_hist_quan
```



```{r, create_obs_table, include = T}
dfp_obs_count <- dft_count_all %>%
  select(
    Regimens = n_regimens,
    `MedOnc Notes` = n_med_onc,
    `CT` = n_scan_ct,
    `MRI` = n_scan_mri,
    `Bone Scans` = n_scan_bone,
    `PET or PET-CT` = n_scan_pet_or_pet_ct,
    `Total scans` = n_scan_total
  ) %>% 
  gtsummary::tbl_summary(data = ., digits = everything() ~ 0) %>%
  modify_header(
    label = '**Observation Type**'
  )
  
```


### R1.4c Observation counts (table)

This is a simpler presentation of the same three statistics as the previous figure:

```{r, include = T}
dfp_obs_count 
```






```{r}
gg_count_hist <- ggplot(data = dft_count_long,
         aes(x = n)) + 
  stat_ecdf(color = '#4477AA') + 
  geom_text_repel(data = dft_count_quantiles,
             # yes, this is screwed up:
             aes(x = y, y = n, label = y),
             hjust = 0, nudge_x = 8, size = 2.5,
             direction = "x",
             segment.size = 0.15
            ) + 
  theme_bw() +
  facet_wrap(vars(type_f), scales = "free", nrow = 2) + 
  scale_y_continuous(
    name = "Cohort fraction",
    breaks = c(0, .5, 1),
    labels = paste0(c(0, .5, 1)*100, "%")
  ) + 
  theme(
    strip.text = element_text(hjust =0)
  )
```












```{r}
# Sunburst plots are created in a distinct script,
#   save_sunbursts.R.
js_full_sun <- readRDS(
  here("data", "sunburst_plots","full.rds")
)
```

### R1.5a Regimens sunburst, full cohort

This is a sunburst plot, which shows the regimen sequences used by participants in the cohort.  The innermost ring shows the first line therapy, the second ring shows second line therapy, and so on.  For tractability I capped the number of regimens at 3, but we can extend further if needed.

This is an **interactive plot**, so you can mouse over each sector to reveal the drugs represented.

*Credit:* MSK stats team for pioneering the use of these plots for drug regimens.

```{r, include = T}
js_full_sun
```



```{r}
js_crpc_sun <- readRDS(
  here("data", "sunburst_plots","crpc.rds")
)
```

### R1.5b Regimens sunburst, post-CRPC

Same plot as previous, limited to regimens started on or after participants were declared Castrate-resistant by medical oncologist note.

**Observation:** Shockingly similar.

```{r, include = T}
js_crpc_sun
```





```{r}
dft_drug <- readr::read_csv(here('data', 'drug.csv'))

dfp_top_drugs <- dft_drug %>%
  group_by(record_id, drug) %>%
  summarize(observed = 1, .groups = "drop") %>%
  group_by(drug) %>%
  summarize(n = sum(observed)) %>%
  arrange(desc(n)) %>%
  head(10)

dc_wrap <- function(drug_name) {
  get_drug_compliments(dft_drug, drug_name) %>%
    reshape_drug_compliments()
}

dfp_top_drugs <- dfp_top_drugs %>%
  mutate(df_top_comp = purrr::pmap(
    .l = list(drug_name = drug),
    .f = dc_wrap
  )) %>%
  tidyr::unnest(df_top_comp) %>%
  mutate(drug = str_replace(drug, "\\(.*\\)", ""))
```


### R1.5c Drugs used, table

The number of participants (n) who used a particular drug in any regimen.  `Compliment 1` states the drug mostly commonly used in regimens containing `Drug`.  

*Example interpretation of a row*  Bicalutamide was used by 627 participants over the course of followup.  Regimens containing bicalutamide also frequently contained Docetaxel (4.1% of regimens with bicalutamide had docetaxel) and Degarelix.

```{r, print_drug_table, include = T}
dfp_top_drugs %>% 
  select(Drug = drug, 
         n = n, 
         `Compliment 1` = first, 
         `Compliment 2` = second) %>%
  huxtable::hux(.) %>%
  huxtable::theme_striped(.)
```









```{r, create_top_drug_plot}
n_prca_cohort <- dft_ca_ind %>% nrow()

gg_top_drugs <- plot_drug_prop(
  drug_dat = dfp_top_drugs,
  cohort_n = n_prca_cohort,
  sample(viridisLite::magma(n = 10, begin = 0.2, end = 0.8)),
  plot_title = glue("Drugs started any time (n={n_prca_cohort})")
)


# Now repeat the calculation above for CRPC regimens only.
dft_cast_block <- make_cast_status_block(
  dft_med_onc, 
  dft_ca_ind, 
  remove_hspc_after_crpc = T
)
n_prca_crpc <- dft_cast_block %>% 
  filter(md_cast_status_f %in% "Castrate-Resistant") %>%
  nrow

dft_drug_crpc <- dft_cast_block %>%
  filter(md_cast_status_f %in% "Castrate-Resistant") %>%
  mutate(crpc_start_days = dx_block_start * 365.25) %>%
  select(record_id, crpc_start_days) %>%
  inner_join(
    dft_drug, 
    ., 
    by = "record_id",
    relationship = "many-to-one"
  )

dft_drug_crpc %<>% 
  # dx_drug_start_int is the number of days from dx to start of drug.
  filter(dx_drug_start_int >= crpc_start_days)

# Same processing from here:
dfp_top_drugs_crpc <- dft_drug_crpc %>%
  group_by(record_id, drug) %>%
  summarize(observed = 1, .groups = "drop") %>%
  group_by(drug) %>%
  summarize(n = sum(observed)) %>%
  arrange(desc(n)) %>%
  head(10) %>%
  mutate(drug = str_replace(drug, "\\(.*\\)", ""))

gg_top_drugs_crpc <- plot_drug_prop(
  drug_dat = dfp_top_drugs_crpc,
  cohort_n = n_prca_crpc,
  pal = sample(viridisLite::viridis(n = 10, begin = 0.2, end = 0.8)),
  plot_title = glue("Drugs started in CRPC setting (n={n_prca_crpc})")
)

gg_top_drug_comb <- cowplot::plot_grid(
  gg_top_drugs, gg_top_drugs_crpc,
  ncol = 1
)
    
```

### *R1.5d Drugs used (All + post-CRPC)

Proportion of the cohort ever exposed to each agent, limited to the top 10 most used.  Top figure shows the entire cohort, bottom figure restricts to drugs started after a CRPC med onc note.  Participants who never had a CRPC note are obviously excluded.

```{r, output_top_drug_plot, include = T, fig.height = 6, fig.width = 7}
gg_top_drug_comb
```















```{r}
js_aed <- readRDS(
  here("data", "sunburst_plots","abi_enza_doce.rds")
)
js_aed_no_dupes <- readRDS(
  here("data", "sunburst_plots","abi_enza_doce_no_dupes.rds")
)

js_rad_doce_no_dupes <- readRDS(
  here("data", "sunburst_plots","radium_doce_no_dupes.rds")
)
```

### *R1.5e Drugs: Abi, Enza, Doce (all regimens)

The group was interested in the order of drugs used for participants who used Abiraterone, Enzalutamide or Docetaxel. The cohort we examine here is anyone who used any of these drugs.  

In the following plot we show the full sequence of regimens containing either of these agent categories:

```{r, include = T}
js_aed
```


**Observations:** 

- Abiraterone Acetate seems to be the most common first line therapy, but all three are used frequently.


### *R1.5f Drugs: Abi, Enza, Doce (repeats removed)

This plot is the same as the previous, except that regimens with repeat enza/abi/doce information are removed.  

For example, if a participant's regimen order was "Abiraterone, Abiraterone, Docetaxel" we simplify this to "Abiraterone, Docetaxel".  But if it was "Abiraterone, Docetaxel, Abiraterone" then we keep the full detail (no repeats exist).


```{r, include = T}
js_aed_no_dupes
```








### *R1.5g Drugs: Radium, Docetaxel (repeats removed)

This plot uses the "repeats removed" format for drug regimens, see R1.5f for more information.  In this our cohort is formed by participants who used either Radium 223 and Docetaxel at any point in their monitoring.


```{r, include = T}
js_rad_doce_no_dupes 
```

**Observations:** 

- Radium is by far less common as a first choice.  Most people who used docetaxel did not go on to take radium.
- Among those who used both Radium 223 and Docetaxel, both orderings were common (12.3% Rad->Doce vs 14.9% Doce->Rad).








```{r}
set.seed(189)
random_pt_subset <- dft_pt$record_id %>% 
  unique %>% 
  sample(x = ., size = 40, replace = F)

dft_cast_block <- make_cast_status_block(
  dft_med_onc, 
  dft_ca_ind, 
  remove_hspc_after_crpc = T)

dft_event <- make_event_df(
  ca_ind_dat = dft_ca_ind,
  cast_block_dat = dft_cast_block
)

gg_swim_cast_dx <- plot_swim_block_status(
  #line_pal = c("gray80", "#ddaa33", "#bb5566"),
  block_dat = dft_cast_block,
  block_var = "md_cast_status_f",
  event_dat = dft_event,
  event_var = "event",
  pt_subset = random_pt_subset
)

ggsave(filename = here("fig", "R1.5a_swim_cast_dx.png"),
       plot = gg_swim_cast_dx,
       width = 7, height = 5, dpi = k_dpi)
```

### R1.6a: Swimmer plot, CRPC/HSPC, from dx

**General note for swimmer plots** For the moment we're using a random subset of participants.  When we decide on a final set of plots for the manuscript, it may make sense to choose a representative cohort based on what we're showing.

We have removed indications of hormone-sensitive prostate cancer in the data which occur after designations of castrate-resistant prostate cancer.  This is based on personal communications with our clinical collaborators, who shared that once someone is considered castrate-sensitive they retain that status for life.

```{r, include = T}
gg_swim_cast_dx
```










```{r}
dft_lab <- dft_ca_ind %>%
  select(record_id, stage_dx)

gg_swim_cast_dx_lab <- plot_swim_block_status(
  #line_pal = c("gray80", "#ddaa33", "#bb5566"),
  block_dat = dft_cast_block,
  block_var = "md_cast_status_f",
  event_dat = dft_event,
  event_var = "event",
  label_dat = dft_lab,
  label_var = "stage_dx",
  order_by_label = T,
  pt_subset = random_pt_subset
)

ggsave(filename = here("fig", "R1.5a_swim_cast_dx_lab.png"),
       plot = gg_swim_cast_dx_lab,
       width = 7, height = 5, dpi = k_dpi)
```

### R1.6b: Swimmer plot, CRPC/HSPC, dx, stage grouping

```{r, swim_crpc_stage_group_output,  include= T}
gg_swim_cast_dx_lab
```









```{r}
dft_cast_block_by_age <- left_join(
  dft_cast_block,
  (dft_ca_ind %>% select(record_id, age_dx)),
  by = "record_id"
) %>%
  mutate(
    across(
      .cols = c(dx_block_start, dx_block_end, tt_os_dx_yrs),
      .fn = (function(x) x + age_dx)
    )
  ) 

dft_event_by_age <- left_join(
  dft_event,
  (dft_ca_ind %>% select(record_id, age_dx)),
  by = "record_id"
) %>%
  mutate(t_yrs = t_yrs + age_dx)

gg_swim_cast_age <- plot_swim_block_status(
  block_dat = dft_cast_block_by_age,
  block_var = "md_cast_status_f",
  event_dat = dft_event_by_age,
  event_var = "event",
  pt_subset = random_pt_subset
)

gg_swim_cast_age <- gg_swim_cast_age + 
  scale_x_continuous(name = "Age (years)")

ggsave(filename = here("fig", "R1.5b_swim_cast_age.png"),
       plot = gg_swim_cast_age,
       width = 7, height = 5, dpi = k_dpi)

```

### R1.6c: Swimmer plot, CRPC/HSPC, by Age

```{r, include = T}
gg_swim_cast_age 
```









```{r}
dft_dmet_block <- make_dmet_status_block(ca_ind_dat = dft_ca_ind)

gg_swim_dmet_dx <- plot_swim_block_status(
  #line_pal = c("gray80", "#ddaa33", "#bb5566"),
  block_dat = dft_dmet_block,
  block_var = "dmet_status",
  event_dat = dft_event,
  event_var = "event",
  pt_subset = random_pt_subset,
  line_pal = c("gray80", "#997700")
)

ggsave(filename = here("fig", "R1.5d_swim_dmet_dx.png"),
       plot = gg_swim_dmet_dx,
       width = 7, height = 5, dpi = k_dpi)
```

### R1.6d: Swimmer plot, Distant metastasis status, from dx

```{r, print_swim_dmet_dx, include = T}
gg_swim_dmet_dx
```










```{r, create_swim_by_site}
set.seed(1898)
random_pt_subset_strat_site <- dft_pt %>%
  group_by(institution) %>%
  sample_n(size = 10) %>%
  ungroup() %>%
  pull(record_id)

dft_cast_block_site <- dft_cast_block %>% 
  left_join(
    .,
    select(dft_pt, record_id, institution),
    by = 'record_id'
  )

dft_event_site <- dft_event %>% 
  left_join(
    .,
    select(dft_pt, record_id, institution),
    by = 'record_id'
  )

gg_swim_cast_dx_site <- plot_swim_block_status(
  #line_pal = c("gray80", "#ddaa33", "#bb5566"),
  block_dat = dft_cast_block_site,
  block_var = "md_cast_status_f",
  event_dat = dft_event_site,
  event_var = "event",
  pt_subset = random_pt_subset_strat_site,
)

gg_swim_cast_dx_site <- gg_swim_cast_dx_site +
  facet_wrap(
    vars(institution), 
    scales = "free_y",
    ncol = 1
  )



ggsave(
  filename = here(
    "fig", 
    "R1.5e_swim_cast_dx_site.png"
  ),
  plot = gg_swim_cast_dx_site,
  width = 7, height = 7, dpi = k_dpi
)
```

### R1.6e: Swimmer plot, CRPC/HRPC by site

This is a different sample of patients than the previous plots - here we select 10 participants per site (stratified random sample).

**Observation:** There are some major differences in the way sites appear to curate CRPC/HSPC.  UHN probably has different curation practices since all patients go from unknown to castrate-resistant.

```{r, print_swim_dmet_dx_site, include = T, fig.height = 7}
gg_swim_cast_dx_site
```





## Todo list

- Clinical characteristics:
  - Add clinical Gleason score to diagnosis table with a footnote.
  - Trim the existing report content once we have the clinical characteristics section settled.
- Genetic reporting:
  - Paused this wihle we answer an important question:  **Do we want to filter mutation data using OncoKB in this cohort?**  This was done in the breast cancer group, and I think it makes sense.
  - Items to report once filtering decided:
    - For each gene, report the proportion of participants tested with a panel that covers that gene.
    - For participants with paired metastatic/primary samples, compare the panel results.
- Survival analyses:
  - Test dependence of genetic testing and death times.  Repeat indexing from metastatic disease.
  - Estimate overall survival using (1) simple risk set adjustment, assuming independent truncation, (2) transformation models allowing for dependent truncation.  Repeat indexing from metastatic disease.
- Clinico-genomic analyses:
  - Goal:  exploring genomic variables which are prognostic/predictive for survival.
  - LTRC Cox model with a LASSO penalty - bootstrap for assessment of how reliable these results are.
  - This can be repeated for other index times of interest.  For example, indexing from first use of Abiraterone Acetate in those with metastatic disease.
  - PFS is challenging but we can start with descriptive work to assess whether it's valuable.

## Appendices

### Testosterone/PSA unit conversions

Testosterone and PSA were recorded under a number of units.  We assumed the following conversion factors:

**PSA** 

- 1 ng/dL is equal to 0.01 ng/mL.
- 1 μg/dL is equal to 1 ng/mL.
- PSA levels in nmol/L (2 results) were discarded as it seemed exceptionally rare to state PSA this way (but common for testosterone, so an easy mistake to make).
- Any PSA results with missing units were discarded (n = `r pull(filter(tm_missing_units, tm_type %in% "PSA"), n)`)

**Testosterone** 

- 1 ng/mL is equal to 100 ng/dL.
- 1 nmol/L is equal to 28.81 ng/dL.
- 1 pg/mL is equal to 0.1 ng/dL.
- 1 μg/dL is equal to 100 ng/dL.




## Problems/Concerns

### *(Kehl) CRPC/HRPC never declared

A point from Ken Kehl and others on 8 May 2023:  For people who were never declared as CRPC or HSPC may never have seen a medical oncologist.  Can we look at stage or subsequent development of distant metastasis to see if that's the case?

The following table shows stage at diagnosis for those who were never declared as hormone sensitive or castrate resistant.  This includes those with zero medical oncologist notes:

```{r}
vec_never_declared <- dfp_post_baseline %>%
  filter(
    `Castration status (first)` %in% c(
      "Never stated in Impression/Plan",
      "Zero med onc notes"
    ) 
  ) %>%
  pull(record_id)
```


```{r, include = T}
dft_ca_ind %>% 
  filter(record_id %in% vec_never_declared) %>%
  tabyl(stage_dx) %>%
  select(-valid_percent) %>%
  filter(!is.na(stage_dx)) %>%
  rename(proportion = percent)
```

**Observation:** There is a lower proportion of stage IV participants in this group compared with the overall cohort (21% vs 38%).  However, there are still a large number of participants with late stage disease at diagnosis which were not designated HSPC/CRPC.

The following table shows whether a distant metastasis was ever noted among that same participant group:

```{r, include = T}
dft_dmet_block %>%
  filter(record_id %in% vec_never_declared) %>%
  group_by(record_id) %>%
  summarize(
    dmet_noted = case_when(
      any(dmet_status %in% "Distant Metastasis") ~ "Yes",
      T ~ "No"
    ),
    .groups = "drop"
  ) %>%
  count(dmet_noted)
```

**Observation:**  Almost all dmets noted come from distant metastasis noted at diagnosis (stage IV = 96, dmets at any time = 103).

### Missing marker units on some results

There are quite a few testosterone/PSA rows with no units.  These have just been excluded form the analysis.  The numbers of exclusions are low relative to the amount of data:

```{r, echo = T, include = T}
dft_tm_miss_units <- dft_tm %>%
  filter(tm_type %in% c("PSA", "Testosterone")) %>%
  filter(is.na(tm_result_units))
dft_tm_miss_units %>%
  mutate(num_result_exists = !is.na(tm_num_result)) %>%
  summarize(
    n_num_result_exists = sum(num_result_exists),
    prop_num_result_exists = mean(num_result_exists)
  )
```




```{r, echo = F, include = F}

# This block can be deleted later on.  Output to resolve a potential error in data.

# dft_tm_miss_units %>%
#   arrange(record_id) %>%
#   readr::write_csv(
#     x = ., 
#     file = here("data", "tumor_marker_missing_units.csv"),
#     na = ""
#   )
# 
# library(synapser)
# 
# synLogin()
# synapser::File(here("data", "tumor_marker_missing_units.csv"),
#                parent = "syn51393619") %>%
#   synStore()


```






