---
title: "BPC Prostate - Genomic descriptives"
author: "Alex Paynter"
date: "`r Sys.Date()`"
output:
  quarto::quarto_html:
    toc: true
    toc_float: true
    theme:
      name: paper
editor_options:
  quarto:
    chunk_output_type: console
execute:
  echo: false
  include: false
  warning: false
  message: false
  fig.width: 7
  fig.height: 5
---

```{r}
library(purrr); library(here); library(fs)
purrr::walk(.x = fs::dir_ls(here('R')), .f = source) # also loads lots of packages.
```

```{r, set_gtsummary_theme}
theme_gtsummary_compact()
theme_gtsummary_language("en", big.mark = "") # for now.
```

```{r}
read_wrap_clin <- function(p) {
  read_rds(file = here("data", 'clin', p))
}

dft_pt <- read_wrap_clin("dft_pt.rds")
dft_ca_ind <- read_wrap_clin("dft_ca_ind.rds")
# we're taking the augmented version, which has TMB columns added.  The existing info is the same.
dft_cpt <- read_wrap_clin("dft_cpt_aug.rds")
```


## Introduction

This document contains analyses related to survival for the GENIE BPC Prostate Cancer Landscape Manuscript group.  Drafting of started in Feb 2024, with brainstorming on analyses of interest in 2023 as well.

The appendix contains a few survival terms we will need for this work (delayed entry, independent left truncation, etc).

## Basics

```{r}

dft_first_cpt <- get_first_cpt(
  ca_ind_dat = dft_ca_ind,
  cpt_dat = dft_cpt
)

dft_surv_dx <- dft_ca_ind %>%
  left_join(., dft_first_cpt, by = c("record_id", "ca_seq")) %>%
  select(
    record_id, ca_seq, stage_dx_iv, dx_cpt_rep_yrs,
    os_dx_status, tt_os_dx_yrs
  ) 

dft_surv_dx %<>% 
  remove_trunc_gte_event(
    trunc_var = 'dx_cpt_rep_yrs',
    event_var = 'tt_os_dx_yrs'
  )

surv_obj_os_dx <- with(
  dft_surv_dx,
  Surv(
    time = dx_cpt_rep_yrs,
    time2 = tt_os_dx_yrs,
    event = os_dx_status
  )
)

gg_os_dx_stage <- plot_one_survfit(
  dat = dft_surv_dx,
  surv_form = surv_obj_os_dx ~ stage_dx_iv,
  plot_title = "OS from diagnosis",
  plot_subtitle = "Adjusted for (independent) delayed entry"
)


```


### R3.1.1 Survival from diagnosis

The following plot shows simple Kaplan-Meier estimates from diagnosis for overall survival.  This is broken into groups for Stage I-III and Stage IV at diagnosis.

```{r}
#| include: true

gg_os_dx_stage
```

Notes:

- The counts for At Risk are snapshots in time.  The counts for Censored and Events are cumulative.
- The at risk set for both groups increases after 0 years due to adjustment for delayed entry.  Participants enter the risk set when they enter the cohort, which can be well after diagnosis.



```{r}
# 
# dft_surv_met <- dft_ca_ind %>%
#   filter(!is.na(os_adv_status)) %>%
#   left_join(., dft_first_cpt, by = c("record_id", "ca_seq")) %>%
#   select(
#     record_id, ca_seq, dx_cpt_rep_yrs,
#     os_adv_status, tt_os_adv_yrs
#   )
# 
# dft_met_timing <- make_dmet_status_block(ca_ind_dat = dft_ca_ind) %>% 
#   filter(dmet_status %in% "Distant Metastasis") %>%
#   select(record_id, ca_seq, dx_met_yrs = dx_block_start)
# 
# dft_surv_met <- left_join(dft_surv_met,
#                           dft_met_timing,
#                           by = c('record_id', 'ca_seq'))
# 
# dft_surv_met %>% 
#   remove_trunc_gte_event(
#     trunc_var = 'dx_cpt_rep_yrs',
#     event_var = 'tt_os_adv_yrs'
#   )
# 
# surv_obj_os_dx <- with(
#   dft_surv_dx,
#   Surv(
#     time = dx_cpt_rep_yrs,
#     time2 = tt_os_dx_yrs,
#     event = os_dx_status
#   )
# )
# 
# gg_os_dx_stage <- plot_one_survfit(
#   dat = dft_surv_dx,
#   surv_form = surv_obj_os ~ stage_dx_iv,
#   plot_title = "OS from diagnosis",
#   plot_subtitle = "Adjusted for (independent) delayed entry"
# )


```


### R3.1.2 Survival from metastasis





## Specific questions




```{r}
dft_met_hrd_surv <- readr::read_rds(
  here('data', 'outcome', 'surv_met_hrd', 'met_hrd_surv.rds')
)

dft_met_hrd_surv %<>% 
  remove_trunc_gte_event(
    trunc_var = 'fmr_fcpt_yrs',
    event_var = 'tt_os_first_met_reg_yrs'
  )

surv_obj_os_fmr <- with(
  dft_met_hrd_surv,
  Surv(
    time = fmr_fcpt_yrs,
    time2 = tt_os_first_met_reg_yrs,
    event = os_first_met_reg_status
  )
)

gg_os_fmr_hrd <- plot_one_survfit(
  dat = dft_met_hrd_surv,
  surv_form = surv_obj_os_fmr ~ hrd_before_entry,
  plot_title = "OS from first metastatic regimen",
  plot_subtitle = "Adjusted for (independent) delayed entry"
)

gg_os_fmr_hrd
  
```




### R3.2.1 HRD pathway

In our Feb 2024 meeting the group expressed an interest in looking at overall survival comparing those with an oncogenic homologous repair deficiency (HRD) mutation, compared with those who do not.  We also discussed that this is complicated by the known lower rates of BRCA2 reported at sites with normal-matched genomic samples, compared with those doing tumor-only sequencing.

**Some more writing here...**

```{r}
#| include: true
gg_os_fmr_hrd
```



Restrictions on this analysis:

- Restrict patient cohort to those with a distant metastasis diagnosed, and who had at least one regimen after this diagnosis.
- 


- Only include genomic tests where the specimen was taken before the first regimen in the metastatic state.
- 
- Only








## Appendix

### Post-death sequencing

We assume that participants enter the GENIE cohort when they have an NGS test **reported**, and use the variable `dx_cpt_rep_yrs` as the marker of cohort entry.  Because it is possible for patients to die before their report is returned, some patients are excluded from survival analyses by this choice.  In this cohort the numbers excluded are not trivial.  

For example, in an analysis of overall survival adjusting for delayed entry in the whole cohort (n=1116), 30 subjects are excluded for this reason.  These participants have much shorter than the included people, so this probably biases our survival estimates high.

Comments on other variables that could be used:

- Aside from report date, there are other times measured.  
- `cpt_order_int`, the interval to the order for the cancer panel test, is unfortunately not complete (especially for UHN), and `dx_path_proc_cpt_yrs`, which is complete but does not seem relevant to cohort entry time.


### Terminology

The following definitions are intentionally narrowed to the scope we will use them in for this work:

- **Index event** or **"Time 0"** - The event we're interested in measuring time from.  Common choices are diagnosis, metastasis, or start of a drug.
- **Event of interest** - Even that we're measuring time to, for example death or progression.
- **Censoring** - When a patient is lost before they have the event of interest occur.  For example, the followup period ended before they died, or they left for another country before a progression occurred.
- **Delayed entry** - A key requirement for recruitment into the BPC cohorts is having a next generation sequencing (NGS, equivalent to "cancer panel test") test.  After recruitment, medical records are retrospectively assessed.  This means that if we're indexing from an event that could occur before NGS (like diagnosis or metastasis), participants can enter the cohort after the index event happens.  Having a gap between index time and cohort entry is known as delayed (cohort) entry.  The reason this is important is that some people may die before they enter the cohort, so delayed entry will introduce selection bias into our recruitment toward those that live longer.
- **Left truncation** - For our purposes this will be equivalent to delayed entry.  The sample is "truncated" because we don't observe those who died before entering the cohort, even if they were diagnosed with the index event.
  - Truncation occurs with patients who we never observe (never entered the cohort).  This differs from censoring happens with patients who entered the cohort, but that we didn't happen to observe the event of interest for.
  - cBioPortal does not currently adjust for left truncation at all, so don't be surprised if the curve there differs dramatically from one here.
  - **Independent left truncation** - The distribution for truncation and the time to event are independent.  In our context would mean, for example, that the time it takes someone to get sequenced is not related to how long they live.  This is obviously not always true, and this has been pointed out with GENIE data specifically: https://jamanetwork.com/journals/jamanetworkopen/fullarticle/2766816.
    - Methods that "adjust for left truncation" usually do so assuming independent left truncation, including most of the plots below.


